### ISSUE TITLE
inconsistent behavior Select Widged

### ISSUE THREAD

**Issue by dominikhoebert** [Textualize/textual#2557]
Getting very inconsistent behavior with the new Select Widged when using their .set_options. Often after the options stay empty, depending on if the Select was used before, an item was selected before or not, or the options before where empy ([]).

```python
from textual import on
from textual.app import App, ComposeResult
from textual.widgets import Select, Button


class SelectBug(App):

    def compose(self) -> ComposeResult:
        self.select = Select(options=[("empty", 0)])
        yield self.select
        yield Button("Change Options", id="change_options")

    @on(Button.Pressed, "#change_options")
    def change_options(self, event):
        self.select.set_options([("Four", 4), ("Five", 5), ("Six", 6)])


if __name__ == "__main__":
    app = SelectBug()
    app.run()
```
Keep the great work up!

**Comment by willmcgugan** [comment]
@davep Suspect this is an issue with `virtual_size` in OptionList (but far from certain).

**Comment by willmcgugan** [comment]
The following diff fixes it. AFAICT.

```
@@ -631,7 +631,7 @@ class OptionList(ScrollView, can_focus=True):
         self.highlighted = None
         self._mouse_hovering_over = None
         self.virtual_size = Size(self.scrollable_content_region.width, 0)
-        self.refresh()
+        self._request_content_tracking_refresh()
         return self

     def _set_option_disabled(self, index: int, disabled: bool) -> Self:
```

If this is the fix, it just needs tests.

**Comment by davep** [comment]
Yup, looks good here.

**Comment by willmcgugan** [comment]
@rodrigogiraoserrao I think Dave is doing this.

**Comment by davep** [comment]
While looking at adding some useful testing for this, I've realised that [the suggested fix](https://github.com/Textualize/textual/issues/2557#issuecomment-1546883815), while it works, is slightly puzzling.

The change sets up a call to `_refresh_content_tracking` via `_on_idle`, which is fine; but beforehand in the code there's already a forced call to `_refresh_content_tracking`. On the surface this would seem to be the same thing, albeit a delayed refresh of the content tracking, rather than an immediate one.

**Comment by willmcgugan** [comment]
You could compare what happens if you remove `overlay: screen` from the select.

**Comment by davep** [comment]
@willmcgugan Good call! Okay, it's not that. And now I look at it it's possibly more obvious: I suspect the problem is that the clear/add is happening while it's `display: none`, which presumably means it has no size at that point.

**Comment by davep** [comment]
Not directly related, but while trying to work out why the `SelectOverlay` doesn't seem to get a `Show` event, I found that `Hide` never seems to happen: #2574.
